lib/utils/class-alerts.php

```diff
@@ -0,0 +1,300 @@
+		}
+
+		return trim( $message );
+	}
+
+	/**
+	 * Validate Opsgenie details array
+	 *
+	 * @param $details array
+	 *
+	 * @return array|WP_Error
+	 */
+	private function validate_details( $details) {
+		$required_keys = [ 'alias', 'description', 'entity', 'priority', 'source' ];
+
+		if ( ! is_array( $details ) ) {
+			return new WP_Error( 'invalid-opsgenie-details', "Invalid \$details: Alerts\:\:opsgenie( " . print_r( $details, true ) ." );" );
+		}
+
+		foreach( $details as $key => $value ) {
+			if ( ! in_array( $key, $required_keys ) ) {
+				return new WP_Error( 'invalid-opsgenie-details', "Invalid \$details: Alerts\:\:opsgenie( " . print_r( $details, true ) ." );" );
+			}
+
+			if ( ! $value ) {
+				return new WP_Error( 'invalid-opsgenie-details', "Invalid \$details: Alerts\:\:opsgenie( " . print_r( $details, true ) ." );" );
+			}
+		}
+
+		return $details;
+	}
+
+	/**
+	 * Get an instance of this Alerts class
+	 *
+	 * @return Alerts|WP_Error
+	 */
+	public static function instance() {
+		if ( null === self::$instance ) {
+			$alerts = new Alerts();
+
+			if ( ! defined( 'ALERT_SERVICE_ADDRESS' ) || ! ALERT_SERVICE_ADDRESS ) {
+				return new WP_Error( 'missing-service-address', 'Missing alerts service host configuration in ALERT_SERVICE_ADDRESS constant' );
+			}
+
+			if ( ! defined( 'ALERT_SERVICE_PORT' ) || ! ALERT_SERVICE_PORT ) {
+				return new WP_Error( 'missing-service-port', 'Missing alerts service port configuration in ALERT_SERVICE_PORT constant' );
+			}
+
+			$alerts->service_address = ALERT_SERVICE_ADDRESS;
+			$alerts->service_port = ALERT_SERVICE_PORT;
+			$alerts->service_url = sprintf( 'http://%s:%s/v1.0/alerts', $alerts->service_address, $alerts->service_port );
+
+			self::$instance = $alerts;
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Send a message to IRC
+	 *
+	 * $level can be an int of one of the following
+	 * NONE = 0
+	 * WARNING = 1
+	 * ALERT = 2
+	 * CRITICAL = 3
+	 * RECOVERY = 4
+	 * INFORMATION = 5
+	 * SCALE = 6
+	 *
+	 * Example Usage
+	 *
+	 * Alerts::chat( '@testuser', 'test message' );			// send testuser a pm on IRC from "a8c"
+	 * Alerts::chat( '@testuser', 'test message', 3 );	// send testuser a pm on IRC with level 'critical'
+	 * Alerts::chat( 'testing', 'test message' );				// have "a8c" join #testing and say something
+	 * Alerts::chat( 'testing', 'test message', 4 );		// have "a8c-test" join #testing and say something with level 'recovery'
+	 *
+	 * @param $target (string) Channel or Username.  Usernames prefixed with an @, channel optionally prefixed by #.
+	 * @param $message (string) Message
+	 * @param $level (int) Level The severity level of the message
+	 * @param $kind string Cache slug
+	 * @param $interval integer Interval in seconds between two messages sent from one DC
+	 *
+	 * @return bool	True if successful. Else, will return false
+	 */
+	public static function chat( $channel_or_user, $message, $level = 0, $kind = '', $interval = 0 ) {
```

Comment:
This is coming from existing code, but I'm not sure why it defaults the `$interval` to `0`...I think we need to change that for this new class...to at least `1`. There is no situation where we need to allow more than 1 message per second of any type.

Reply:
The `$kind` defaulting to an empty string is also weird...it means that the rate limiting won't work (even if interval was set) by default, which to me feels pretty broken. We don't necessarily need to fix it in this PR, but a a default `$kind` that was the MD5 of the `$channel_or_user . $message` would at least provide some safeguards to runaway code.

Reply:
I think the idea of the original code was to make it optional. Perhaps to allow some alerts to skip rate limiting? I agree with the defaults for `$interval` though. Setting it to `0` makes no sense.

Reply:
I believe it's like that so the site itself can send more than 1 message per second across all message types (if $kind is not specified), which is a little dangerous b/c that's a _lot_ of messages from one site. The alerts service endpoint does its own rate limiting so we're not directly at risk of maxing out IRC or Slack, but it would be possible to overwhelm the rate limit service if runaway code was blasting alerts at it.

Reply:
I'm ok keeping the rate limiting as-is to get this PR across the line, just thought it would be a good opportunity to revisit it since this is not used anywhere.

Reply:
@nickdaugherty I liked your suggestion so made it so that there's a default `$kind` value making the rate limiting no longer optional. Added in bd2a2b4b87600ea38d87ddcf6beb48b8451f75fa